### PR TITLE
Force Fritz!Fon image refresh of changed/updated images

### LINKED ISSUE #84: Fritz!Fon does not update images when changed image was uploaded to Fritz!Box
(PR will follow!)

Hardware: Fritz!Box (FW 06.85) Fritz!Fon C5 (FW 02.04.09)

Steps to reproduce:
* Create a vCard with an initial image and upload vCard and image (-i option) to FritzBox
* On FritzFon: check contact > details > image
* Now change contacts image in CardDav server
* Repeat download/upload of vCard with -i option (Console log will show `Uploaded/refreshed 1 of xx image file(s)`)
* On FritzFon: check again contact > details > image again

**Expected: new image is shown in FritzFon
Actual: old image is shown in FritzFon**

Observation#1: in the FritzBox web interface the new image is shown
Observation#2: in the FritzBox FTP server the new image has a newer time stamp (so, it actually was uploaded!)

I tried various ways to force the update of the image inside FritzFon
* Waited 24 hrs (=> no image update)
* Switched to a different phonebook and then back to main phone book (=> no image update)
* Rebooted FritzFon  (=> no image update)
* changed my carddav2fb config.php to an include filter that actually does not pass any vCards, resulting in an empty main phonebook at the FritzFon. Then restored old filter to re-upload all former contacts with their images (=> no image update)
* Deleted all files from the FTP USB fonpix folder then re-uploaded all fotos  (=> no image update)

Seems the FritzFon has some very sticky internal cache implemented (that stupidly ignores the FTP servers timestamp of the files) :-(

I then tested a "quick hack": 
* If I rename the image file in FTP dir (e.g., append a date/timestamp *_YYMMDDHHMMSS.jpg)
* and change the imageURL in XML phonebook dump accordingly
* then re-import XML phonebook dump to FritzBox
=> Then FritzFon shows the new image! Hey! Finally! Sigh!
🎉

I prepared a PR that does exactly this on image FTP uploads:
* Append timestamp to all initial uploaded image names
* On following image uploads check if image file size has changed
* If yes: delete the old image file on FTP and upload the new image with a new timestamp
* imageURL calculation now happens in functions.php instead of converter.php

### PR DESCRIPTION
**Pull request by derwok**
by changing image URLs of updated images.
FTP image files and image URLs now get an appended time stamp "_YYMMDDHHMMSS".
Outdated images are deleted from FTP server.
Fixes #84

### CODE REVIEW

File: src/functions.php [hunk 1]
```diff
@@ -63,34 +71,60 @@ function uploadImages(array $vcards, $config, callable $callback=null)
     if (!ftp_login($ftp_conn, $config['user'], $config['password'])) {
         error_log("ERROR: Could not log in ".$config['user']." to ftp server ".$ftpserver." for image upload.");
         return false;
-    } 
+    }
     if (!ftp_chdir($ftp_conn, $config['fonpix'])){
         error_log("ERROR: Could change to dir ".$config['fonpix']." on ftp server ".$ftpserver." for image upload.");
```
Comment (andig) [inline src/functions.php]: hier fehlt ein not

File: src/functions.php [hunk 2]
```diff
@@ -63,34 +71,60 @@ function uploadImages(array $vcards, $config, callable $callback=null)
...
+
+                // Upload new image file
                 $memstream = fopen('php://memory', 'r+');     // we use a fast in-memory file stream
                 fputs($memstream, $vcard->rawPhoto);
                 rewind($memstream);
 
-                // upload file
-                if (ftp_fput($ftp_conn, $remotefilename, $memstream, FTP_BINARY)){
+                // upload new image
+                if (ftp_fput($ftp_conn, $newFTPimage, $memstream, FTP_BINARY)){
```
Comment (andig) [inline src/functions.php]: bitte space vor {

File: src/functions.php [hunk 3]
```diff
@@ -63,34 +71,60 @@ function uploadImages(array $vcards, $config, callable $callback=null)
...
+                $newFTPimage = sprintf('%1$s_%2$s.jpg', $vcard->uid, $timestampPostfix);
+                if (array_key_exists($vcard->uid, $mapFTPUIDtoFTPImageName)) {
+                    $currentFTPimage = $mapFTPUIDtoFTPImageName[$vcard->uid];
+                    if (ftp_size($ftp_conn, $currentFTPimage) == strlen($vcard->rawPhoto)) {
+                        // No upload needed, but store old image URL in vCard
+                        $vcard->imageURL = $configImagepath . $currentFTPimage;
+                        continue;
+                    }
+                    // we already have an old image, but the new image differs in size
+                    ftp_delete ($ftp_conn, $currentFTPimage);
```
Comment (andig) [inline src/functions.php]: bitte space vor ( weg

File: src/functions.php [hunk 4]
```diff
@@ -63,34 +71,60 @@ function uploadImages(array $vcards, $config, callable $callback=null)
...
+    }
     if (!ftp_chdir($ftp_conn, $config['fonpix'])){
         error_log("ERROR: Could change to dir ".$config['fonpix']." on ftp server ".$ftpserver." for image upload.");
         return false;
-    } 
+    }
+
+    // Build up dictionary to look up UID => current FTP image file
+    $ftpFiles = ftp_nlist($ftp_conn, ".");
+    if (!$ftpFiles){
```
[the remark this answers is not in the document]
Comment (andig) [reply src/functions.php]: bitte space vor {

File: src/functions.php [hunk 5]
```diff
@@ -45,12 +45,20 @@ function download(Backend $backend, $substitutes, callable $callback=null): arra
  *
  * @param $vcards     array     downloaded vCards
  * @param $config     array
- * @return            array     number of uploaded/refreshed images; number of total found images
+ * @return            array     false or [number of uploaded images, number of total found images]
```
Comment (andig) [inline src/functions.php]: return type wäre dann mixed statt array

**Comment by derwok** [comment]
Hey, thx for the feedback!
I fixed review comments & resolved merge conflicts with master.

**Comment by andig** [comment]
Much appreciated, thank you!

**Pull request merged**